design-documents/elasticsearch/extended-analyzer-config.md

```diff
@@ -0,0 +1,278 @@
+        <de_DE>stopwords_de_DE.csv</de_DE><!-- de_DE locale stop words -->
+        <!-- elements for other locales -->
+    </stopwords_file>
+</config>
+```
+
+There may be much more use-cases when SI integrators would like to change default index settings to adjust search results.
+
+To allow complex changes of default Elasticsearch configuration in Magento 2, such us required by Japanese localization project, at the current moment we have 2 options:
+
+1. Use [pluginization](https://devdocs.magento.com/guides/v2.2/extension-dev-guide/plugins.html) mechanism on `\Magento\Elasticsearch\Model\Adapter\Index\Config\EsConfigInterface` and `Magento\Elasticsearch\Model\Adapter\Index\BuilderInterface`.
+1. Introduce explicit extension points and extend possibilities of Magento 2 ElasticSearch configuration.
+
+Plugins are powerful mechanism but they push developers to explore the internals of pluginized module that violates the Open-Closed principle. Usage of plugins would require a same code from project-to-project.
+
+## Decision
+
+Magento 2 should provide a possibility to configure analyzer for indexes. To not limit all the power of ElasticSearch without exposing all complexity of configuration should include:
+- tokenizer configuration
+- token filters configuration
+    - customized filters configuration
+    - list of token filters to be used by an analyzer
+- char filters configuration
+    - customized filters configuration
+    - list of token filters to be used by an analyzer
+
+The extended schema should be fully backward compatible and consistent with the current implementation.
+
+### General Concepts
+As all required configuration options are independent they should be expressed as top-level XML elements.
+
+As all required configuration options have a strong dependency on a language and to be consistent with existing configuration elements each introduced elements should contain `default` element and elements with names corresponding to locale names.
+
+As particular tokenizer and filters may require complex types for configuration XML schema should allow to do that and converting of available options described at ElasticSearch Documentation in JSON to XML elements should follow simple straight-forward rules.
+
+### Tokenizer Configuration
+A tokenizer is configured by `tokenizer` element which must include `default` element as a container for common configuration and may include one or more elements with valid locale codes as a name that are containers for locale-specific configuration.
+
+Each direct child of `tokenizer` element must contain `type` element with the name of tokenizer to be used.
+
+Each direct child of `tokenizer` element may contain element which represent configuration parameter available for specified tokenizer type. See [ElasticSearch parameters XML representation](#ElasticSearch-parameters-XML-representation) section for more information.
+
+```xml
```

Comment:
From what I understood the intention was to make configuration similar to what we already have. Current configuration in the module is different from configuration that we have for other modules. Have you thought about unifying it? Since we don't have a lot of configuration, maybe we can introduce new format, deprecate old one and create fallback for backwards compatibility?

I need to think more about alternative format, but here is what I have so far
```xml
<config>
    <languages>
        <language name="default">
            <tokenizer name="default" type="standard"/>
            <token_filter name="default" type="lowercase"/>
            <char_filter name="default" type="html_strip"/>
        </language>
        <language name="jp_JP">
            <tokenizer name="my_tokenizer"
                       type="kuromoji_tokenizer"
                       mode="extended"
                       discard_punctuation="false"
                       user_dictionary="userdict_ja.txt"/>
            <token_filer name="my_token_filter"
                         type="standard"
                         max_token_length="5"/>
            <char_filer name="my_char_filter"
                        type="pattern_replace"
                        pattern="(\\d+)-(?=\\d)"
                        replacement="$1_"/>
        </language>
    </languages>
</config>
```

Main benefits
1. Name of the language is value of an attribute instead of name of the node, more consistent with other Magento configuration. Are there other benefits in proposed configuration except consistency with existing module configuration?
2. Configuration per language defined in one place
3. Nodes that unlikely will have their own attributes made attributes

Don't like how defaults being set in my example, need to think more about it.

Reply:
My main intention was to make configuration similar to existing exactly in `esconfig.xml` regardless it differs from main approach in other config files. I also think that supporting two types of config will create too many difficulties in maintenance and overcomplicate implementation.

I suppose in future we will need BiC in this module to eliminate support of ES 2 and full support of new features. I think that time in future would be ideal for reviewing the whole config.

Reply:
I would favor introducing configuration in the format that follows general and Magento best practices because amount of configuration we going to add is pretty pig. Currently we have small amount of configuration and it should be easy to support fallback to it. Having free form configuration that is not following conventions of other configuration in Magento is bad for developer experience. Would be interested in other people opinion on this.